Bug 1733365 - Document testResetGlean on JavaScript SDK + initialize page restructure (#1973)
mozilla/glean: docs/user/reference/general/initializing.md

# Initializing

Glean needs to be initialized in order to be able to send pings,

> record metrics and perform maintenance tasks. Thus is is advised that Glean be initialized as

Reviewer: ```suggestion
record metrics and perform maintenance tasks. Thus it is advised that Glean be initialized as
```

soon as possible in an application's lifetime and importantly, before any other libraries in the
application start using Glean.

{{#include ../../../shared/blockquote-info.html}}

## Libraries are not required to initialize Glean

> Libraries rely on the same Glean singleton as the application in which they are
> embedded. Hence, they are not expected to initialize Glean
> as the application should already do that.

## Behavior when uninitialized

Any API called before Glean is initialized is queued and applied at initialization.
To avoid unbounded memory growth the queue is bounded (currently to a maximum of 100 tasks),
and further calls are dropped.

The number of calls dropped, if any,
is recorded in the `glean.error.preinit_tasks_overflow` metric.

## Behavior once initialized

### When upload is enabled

> Once initialized, if upload is enabled,
> Glean will automatically start applying enqueued tasks, such as
> collection of [baseline metrics](../../user/pings/metrics.md).

Reviewer: We shouldn't call them "baseline" metrics or people will confuse that with the baseline ping.
I think it's a bit confusing to keep talking about "tasks" here, that's not a concept exposed to users.
Additionally this links to one specific ping. Maybe we can rephrase that a bit.

```suggestion
Once initialized, if upload is enabled,
Glean applies all metric recordings, for both user-defined metrics and builtin metrics.
This always happens off the main thread.
```

(not sure about the last sentencen, but I guess it's good to call it out once?)

Author: This is actually just copy pasted from what is there already, but I agree with the changes. Just need to change the "off the main thread" call, because that does not apply to Javascript. I'll send these changes in a follow-up commit.

### When upload is disabled

> If upload is disabled, any persisted metrics, events and pings (other than `first_run_date` and `first_run_hour`) are cleared, any pending deletion-request is sent and subsequent calls to record metrics and submit pings will be no-ops. Because Glean does that as part of its initialization, users are _required_ to always initialize Glean.

Reviewer: Add some newlines for easier editing and a bit of restructuring.

```suggestion
If upload is disabled, any persisted metrics, events and pings (other than `first_run_date` and `first_run_hour`) are cleared.
Pending `deletion-request` pings are sent.
Subsequent calls to record metrics and submit pings will be no-ops.
Because Glean does that as part of its initialization, users are _required_ to always initialize Glean.
```

> **Glean must be initialized even if upload is disabled**.

Reviewer: ```suggestion
**Glean must be initialized even if upload is disabled.**
```

This does not apply to special builds where telemetry is disabled at build time.
In that case, it is acceptable to not call initialize at all.

## API

### `Glean.initialize(configuration)`

Initializes Glean.

May only be called once. Subsequent calls to `initialize` are no-op.

#### Configuration

The available initialize configuration options may vary depending on the SDK.
Below are listed the configuration options available on most SDKs.

- `applicationId`: Application identifier. For Android and iOS applications, this is the id used on the platform's respective app store and is extracted automatically from the application context.
- `uploadEnabled`: The user preference on whether or not data upload is enabled.

> - `appChannel`: The application's release channel. When present, the `app_channel` will be reported on all pings [`client_info`](../../user/pings/index.html#the-client_info-section) section.

Reviewer: ```suggestion
- `appChannel`: The application's release channel. When present, the `app_channel` will be reported in all ping's [`client_info` section](../../user/pings/index.html#the-client_info-section).
```

- `appBuild`: A build identifier e.g. the build identifier generated by a CI system (e.g. "1234/A"). If not present, `app_build` will be reported as "Unknown" on all pings [`client_info`](../../user/pings/index.html#the-client_info-section) section.
- `appDisplayVersion`: The user visible version string for the application running Glean. If not present, `app_display_version` will be reported as "Unknown" on all pings [`client_info`](../../user/pings/index.html#the-client_info-section) section.
- `serverEndpoint`: The server pings are sent to. Defaults to `https://incoming.telemetry.mozilla.org`.

To learn about SDK specific configuration options available, refer to the [Reference](#reference) section.

{{#include ../../../shared/blockquote-stop.html}}

> ##### Always initialize Glean with the correct value upload preference

Reviewer: ```suggestion
##### Always initialize Glean with the correct upload preference
```

> Glean must **always** be initialized with real values.
>

> > Always pass the user preference, e.g. `Glean.initialize(upload=userSettings.telemetry_enabled)` or the equivalent for your application. Never call `Glean.initialize(upload=true)` if `true` is a placeholder value that later gets reset by `Glean.setUploadEnabled(false)` and vice-versa.

Reviewer: ```suggestion
> Always pass the user preference, e.g. `Glean.initialize(upload=userSettings.telemetry_enabled)` or the equivalent for your application.
> Never call `Glean.initialize(upload=true)` if `true` is a placeholder value that later gets reset by `Glean.setUploadEnabled(false)` and vice-versa.
```

Other test-running libraries have similar features, but are different in the details.

Create a file `conftest.py` at the root of your test directory, and add the following to reset Glean at the start of every test in your suite:

```python
import pytest
from glean import testing

@pytest.fixture(name="reset_glean", scope="function", autouse=True)
def fixture_reset_glean():
    testing.reset_glean(application_id="my-app-id", application_version="0.1.0")
```

</div>

<div data-lang="Rust" class="tab"></div>
<div data-lang="JavaScript" class="tab">

The Glean JavaScript SDK contains a helper function `testResetGlean()` for resetting Glean for tests.
It expects the same list of arguments as `Glean.initialize`.

Each reset of the Glean JavaScript SDK will clear stores. Calling `testResetGlean` will also make
metrics and pings testing APIs available and replace ping uploading with a mock implementation that
does not make real HTTP requests.

```js
import { testResetGlean } from "@mozilla/glean/testing"

describe("myTestSuite", () => {
    beforeEach(async () => {
        await testResetGlean("my-test-id");
    });
});

```

</div>

> <div data-lang="Firefox Desktop" class="tab"></div>

Author: @chutten I am assuming testing Glean metrics in Firefox Desktop means Glean resetting is handled internally, is that correct? Would a message such as "On Firefox Desktop Glean resetting is handled internally." be correct here?

Reviewer 2: Accccctttually, we just implemented `Services.fog.testResetFog()` for test reset capability (cf [bug 1741674](https://bugzilla.mozilla.org/show_bug.cgi?id=1741674)). You can read all about it in [our brand new instrumentation testing docs](https://firefox-source-docs.mozilla.org/toolkit/components/glean/user/instrumentation_tests.html).

What I missed was linking heavily to that doc from the `Firefox Desktop` tabs in all the metric types. Want me to file a bug for that?

Author: Nah, we do not mention resetting for any of the SDKs on the metric types pages. Actually this PR is the first time resetting is documented in the reference part of the book. Before it was mentioned in the "testing metric" guide only.

Author: Back to "what I should add here", maybe the tooltip can say "Resetting in FOG is documented on <this link you just sent me>"? WDYT?

Reviewer 2: Sounds good to me.
